### ISSUE TITLE
rotate molecule with axis = atom1 - atom2 

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#27]
@drroe 

I am not sure how to use "rotate" in cpptraj to do this "Rotate the molecule by an amount (in degrees, specified by -d) along the axis defined by two atoms (specified by -a1 and -a2)" [from antechamber](http://ambermd.org/antechamber/ac.html)

Does cpptraj support this? thanks

cpptraj: https://github.com/Amber-MD/cpptraj/blob/master/src/Action_Rotate.cpp#L9-L10

**Comment by drroe** [comment]
There isn't an action that supports this, but you can do it with the Frame, Vec3, and Matrix_3x3 classes (following example assumes MyFrame already has coordinates):

```C++
// a1 and a2 are integers representing atom indices (from 0)
Vec3 axisOfRotation = MyFrame.SetAxisOfRotation(a1, a2);
// theta is a double, must be in radians
Matrix_3x3 U;
U.CalcRotationMatrix(axisOfRotation, theta);
MyFrame.Rotate( U ); // Rotates all coords
```
Do you need this to be an Action?

**Comment by hainm** [comment]
@drroe I don't need extra action for my need since I can follow above code. But it's great if you can introduce to cpptraj. I think cpptraj has already some actions that can change coordinates, we can add more (for example, I really like Action_Makestructure).

**Comment by drroe** [comment]
@hainm This is something that could be introduced as e.g. a keyword to the Rotate action, but I'm curious as to why you would ever want to do this. Do you have specific applications in mind?

**Comment by hainm** [comment]
@drroe I was thinking about rotating a group of atoms along a defined axis, translating it to a pocket. In the future, we can extend pytraj/parmed/cpptraj for modelling. 

I am playing with force field, so I would like to use `cpptraj` to rotate some dihedral angles then calling `pysander` to evaluate energy. I am too lazy to look at tleap.

**Comment by drroe** [comment]
Does the 'makestructure' action not do what you want? Or have you been
having trouble getting it to work? I'm not able to access my files easily
right now but I can get you some examples tomorrow.

-Dan

PS - Related, I will have to start adding more test cases (like you've
requested) since they do serve as examples.

On Tuesday, August 4, 2015, Hai Nguyen <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> @drroe <https://github.com/drroe> I was thinking about rotating a group
> of atoms along a defined axis, translating it to a pocket. In the future,
> we can extend pytraj/parmed/cpptraj for modelling.
>
> I am playing with force field, so I would like to use cpptraj to rotate
> some dihedral angles then calling pysander to evaluate energy. I am too
> lazy to look at tleap.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Amber-MD/cpptraj/issues/27#issuecomment-127856810>.
>


-- 
-------------------------
Daniel R. Roe, PhD
Department of Medicinal Chemistry
University of Utah
30 South 2000 East, Room 307
Salt Lake City, UT 84112-5820
http://home.chpc.utah.edu/~cheatham/
(801) 587-9652
(801) 585-6208 (Fax)

**Comment by hainm** [comment]
@drroe uhm, I will look at cpptraj's manual again for other cases. My above example (rotate_dihedral) called `Action_Makestructure` for residue 3, omega.

But different examples from you will be great help.

**Comment by drroe** [comment]
OK, so makestructure has a few different modes you can use:
1) Apply a pre-defined structure type to a specified residue range via manipulation of phi/psi dihedrals. The dihedrals are applied either 1 at a time for basic structure (e.g. alpha) or 2 at a time in case of turns. Simple example:
```
parm ../tz2.parm7
trajin ../tz2.nc 1 1
# Apply polyproline II dihedral to residues 1-13
makestructure pp2:1-13
outtraj pp2.rst7
```
2) Apply custom structure type to a specified residue range via manipulation of phi/psi dihedrals.
```
# Make residues 1 and 12 pre-defined "extended"
# Make residues 2-5 turn "custom1"  (phi1=-80, psi1=130, phi2=-130, psi2=140)
# Make residues 6-7 turn type I'
# Make residues 8-11 turn "custom2" (phi1=140, psi1=170, phi2=-100, psi2=140)
makestructure extended:1,12 \
  custom1:2-5:-80.0:130.0:-130.0:140.0 \
  typeI':6-7 \
  custom2:8-11:-140.0:170.0:-100.0:140.0
```
3) Apply specific dihedral value to a specific dihedral type.
```
# Make residue 5 phi 90 degrees
makestructure customdih:5:phi:90
```
4) Apply specific dihedral value to a custom dihedral type.
```
# Make residue 8 dihedral "chi1" (defined by atoms N, CA, CB, CG) 35 degrees
makestructure chi1:8:N:CA:CB:CG:35
```
5) Apply dihedral values from a reference structure to target structure. This one is less well-tested so it may not work as advertised (I need to check that out).

Hopefully this helps. Let me know if you have more questions.

**Comment by drroe** [comment]
Forgot to put an example for 5). It manipulates backbone phi/psi only I think.
```
parm ../tz2.parm7
reference ../tz2.rst7
trajin pp2.rst7.save
# Apply backbone dihedrals from reference structure residues 1-13 to residues 1-13
makestructure "ref:1-13:tz2.rst7"
```

**Comment by hainm** [comment]
Thanks. I will try all of them and will ask you later if there is any issue.

**Comment by hainm** [comment]
@drroe 

In this paper, the authors said `The δ dihedral angle (C5′−C4′−C3′−O3′) was set to either 140° or 81°`. Can `cpptraj` make this change? if yes, can you give me an example? thanks.

http://pubs.acs.org/doi/full/10.1021/ct900604a

**Comment by drroe** [comment]
Since delta is a standard dihedral, you can just use:
```
makestructure delta:5:delta:140
```
Assuming residue 5 has a delta dihedral to rotate. The syntax is a little awkward in that the first field is always a name (in this case the third field is the dihedral type) - I am considering changing it so that it is a keyword instead. I am also considering just making a whole new action specifically for dihedral rotations (as opposed to applying pre-defined phi/psi angles), which may be a little easier to understand conceptually (let me know if you have an opinion on this). Let me know if it doesn't work for you.

**Comment by hainm** [comment]
thanks Dan. It works.

```python
In [14]: pt.dihedral_analysis.calc_delta(traj, resrange='5')
Out[14]: 
<pytraj.DatasetList with 1 datasets>
delta:5
[ 89.73265527  70.59254293  73.93549047]

In [15]: pt.make_structure(traj, 'delta:5:delta:120')

In [16]: pt.dihedral_analysis.calc_delta(traj, resrange='5')
Out[16]: 
<pytraj.DatasetList with 1 datasets>
delta:5
[ 120.00000334  120.00000334  120.00000334]
```

Per syntax, let me think first. @swails might have better idea.

**Comment by hainm** [comment]
ah, I absolute agree that cpptraj should have new action for dihedral rotation. (`MakeStructrure` can do some but its name is not really good for this purpose).

**Comment by drroe** [comment]
Addressed by #247
